nemo_aligner/models/mm/stable_diffusion/image_text_rms.py

```diff
@@ -121,9 +122,147 @@ def model_provider_func(self, pre_process, post_process):
             pre_process=pre_process,
             post_process=post_process,
         )
-
         return model
 
+    def forward(self, images, captions):
+        rewards = self.get_reward(images, captions)
+        return rewards
+
+    def loss_func(self, output_tensor):
+        """ 
+        rewards_0, rewards_1 - tensor of size [B,] each that computes rewards 
+        labels - Bx2 tensor containing labels
+        """
+        rewards_0, rewards_1, labels = output_tensor
+        logits = torch.stack([rewards_0, rewards_1], 1)  # [B, 2]
+        logsoftmax = F.log_softmax(logits, dim=1)
+        labels_s = labels + 1e-4
```

Comment:
is this for stability? if so why 1e-4? The default value is usually 1e-6 in other cases; also add it to the denominator in the next line

Reply:
1e-4 seems to be fine if the labels are just 0, 0.5, and 1.

The denominator already has the 1e-4 added from the previous line (to normalize `label_s` to a valid pdf).